Improve private chains tracking

abi87 opened the pull request:
Along @cronicc indications we are going to extend getchaintips with a --with-penalties=true/false paramenter to include, for each branch:

1. penalty-at-start of branch. A non-zero value indicates branch starts of penalized
2. penalty-at-tip of branch
3. blocks-to-mainchain, i.e. how many blocks on selected branch must be mined in order for it to become the mainchain

abi87 committed:
Minor preparatory refactoring

abi87 committed:
Refactoring to extract logic of block to overran chainActive tip

abi87 committed:
Added extra fields in getchaintips

abi87 committed:
Adding py tests for verbose getchaintips

alsala commented on src/zen/delay.cpp:
```diff
@@ -2,35 +2,31 @@
+    {
+    	// blocks on tip get penalty reduced by 1;
+    	// others get penalty increased proportionally to distance from tip
+    	return std::max<int>((activeChainHeight - newBlock.nHeight),-1);
```
This is more compact than before but we have not the trace anymore; should we keep it?

abi87 committed:
reintroduced trace as for code review indications

albertog78 commented on src/zen/delay.cpp:
```diff
@@ -2,35 +2,35 @@
+    {
+    	// blocks on tip get penalty reduced by 1;
+    	// other blocks get penalty increased proportionally to distance from tip
+    	int64_t blockDelay =  std::max<int64_t>((activeChainHeight - newBlock.nHeight),int64_t(-1));
```
Is this code behaving the same way if for example activeChainHeight is MIN_INT64  and newBlock.height is MIN_INT64 +1? 
Why this PR needed this change?

abi87 replied:
Change is purely to improve readability, I can roll it back should there not be agreement on such improvement.
Moreover this function is called uniquely by `AddToBlockIndex`, with `activeChainHeight ` set to `chainActive.Height()`. `chainActive.Height()` is either positive, or -1 (at start, when no block has been added yet). In both cases, the value far away for int64_t boundaries, which would cause the refactored code to diverge from the original

abi87 replied:
Reverted change as agree offline

abi87 committed:
Reverted unneeded changes to zen/delay.cpp

cronicc committed:
Fix transient failure on MacOS:

In cases where 2 branches have the same length, the 'active' branch
is not guaranteed to be at list index 0 in the getchaintips response.
Fixed by sorting the list on 'status' k:v.
